controller/AuthoringTool.php

```diff
@@ -64,27 +73,78 @@ protected function getValidatedLtiMessagePayload(): LtiMessagePayloadInterface
     }
 
     /**
-     * @throws common_exception_Error
      * @throws ActionEnforcingException
      * @throws InterruptedActionException
+     * @throws LtiException
+     * @throws ContainerExceptionInterface
+     * @throws NotFoundExceptionInterface
+     * @throws common_exception_Error
      */
     public function launch(): void
     {
-        $message = $this->getValidatedLtiMessagePayload();
+        $ltiMessage = $this->getLtiMessageOrRedirectToLogin();
 
         $user = $this->getServiceLocator()
             ->getContainer()
             ->get(tao_models_classes_UserService::class)
             ->addUser(
-                $message->getUserIdentity()->getIdentifier(),
+                $ltiMessage->getUserIdentity()->getIdentifier(),
                 helpers_Random::generateString(UserService::PASSWORD_LENGTH),
-                new core_kernel_classes_Resource(current($message->getRoles()))
+                new core_kernel_classes_Resource(current($ltiMessage->getRoles()))
             );
         $this->getServiceLocator()
             ->getContainer()
             ->get(LtiService::class)
-            ->startLti1p3Session($message, $user);
+            ->startLti1p3Session($ltiMessage, $user);
 
         $this->forward('run', null, null, $_GET);
     }
+
+    /**
+     * @return bool
+     * @throws ContainerExceptionInterface
+     * @throws NotFoundExceptionInterface
+     */
+    private function isFeatureTaoAsToolEnabled(): bool
+    {
+        return $this->getServiceManager()
+            ->getContainer()
+            ->get(FeatureFlagChecker::class)
+            ->isEnabled(FeatureFlagCheckerInterface::FEATURE_FLAG_TAO_AS_A_TOOL);
+    }
+
+    /**
+     * @return LtiMessagePayloadInterface
+     * @throws ContainerExceptionInterface
+     * @throws InterruptedActionException
+     * @throws LtiException
+     * @throws NotFoundExceptionInterface
+     */
+    private function getLtiMessageOrRedirectToLogin(): LtiMessagePayloadInterface
+    {
+        if (!$this->isFeatureTaoAsToolEnabled()) {
+            $this->getLogger()->info(
+                'TAO as tool feature is disabled. The user will be redirected to the login page.'
+            );
+            $this->redirect(_url('login', 'Main', 'tao'));
+        }
+
+        try {
+            $message = $this->getValidatedLtiMessagePayload();
+        } catch (LtiException $exception) {
+            if ($exception->getMessage() !== self::LTI_NO_MATCHING_REGISTRATION_FOUND_MESSAGE) {
```

Comment:
We should use Exception code or Exception type to define this not message string.

Reply:
Sometimes exceptions don't implement the code and its hard to add it, maybe that's the case. But of course better would be to use code of specific exception class to check.

Reply:
Can you make changes to the exception so it will be easy to recognise the type of exception?

Reply:
I would avoid creating a new release on Lti1p3Core only for this.